limiter_atomic_int64.go

```diff
@@ -0,0 +1,94 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ratelimit // import "go.uber.org/ratelimit"
+
+import (
+	"time"
+
+	"sync/atomic"
+)
+
+type atomicInt64Limiter struct {
+	//lint:ignore U1000 Padding is unused but it is crucial to maintain performance
+	// of this rate limiter in case of collocation with other frequently accessed memory.
+	prepadding [64]byte // cache line size = 64; created to avoid false sharing.
+	state      int64    // unix nanoseconds of the next permissions issue.
+	//lint:ignore U1000 like prepadding.
+	postpadding [56]byte // cache line size - state pointer size = 64 - 8; created to avoid false sharing.
+
+	perRequest time.Duration
+	maxSlack   time.Duration
+	clock      Clock
+}
+
+// newAtomicBased returns a new atomic based limiter.
+func newAtomicInt64Based(rate int, opts ...Option) *atomicInt64Limiter {
+	// TODO consider moving config building to the implementation
+	// independent code.
+	config := buildConfig(opts)
+	perRequest := config.per / time.Duration(rate)
+	l := &atomicInt64Limiter{
+		perRequest: perRequest,
+		maxSlack:   time.Duration(config.slack) * perRequest,
+		clock:      config.clock,
+	}
+	atomic.StoreInt64(&l.state, 0)
+	return l
+}
+
+// Take blocks to ensure that the time spent between multiple
+// Take calls is on average time.Second/rate.
+func (t *atomicInt64Limiter) Take() time.Time {
+	var (
+		newTimeOfNextPermissionIssue int64
+		now                          int64
+	)
+	for {
+		now = t.clock.Now().UnixNano()
+		timeOfNextPermissionIssue := atomic.LoadInt64(&t.state)
+		// If this is our first request, then we allow it.
+		if timeOfNextPermissionIssue == 0 {
+			newTimeOfNextPermissionIssue = now
+			if atomic.CompareAndSwapInt64(&t.state, 0, newTimeOfNextPermissionIssue) {
+				break
+			}
+			continue
+		}
+
+		if now-timeOfNextPermissionIssue > int64(t.maxSlack) {
+			// a lot of nanoseconds passed since the last Take call
+			// we will limit max accumulated time to maxSlack
+			newTimeOfNextPermissionIssue = now - int64(t.maxSlack)
+		} else {
+			// calculate the time at which our permission was issued
+			newTimeOfNextPermissionIssue = timeOfNextPermissionIssue + int64(t.perRequest)
+		}
```

Comment:
```suggestion
                 switch {
                 case timeOfNextPermissionIssue == 0:
                      newTimeOfNextPermissionIssue = now
                 case now-timeOfNextPermissionIssue > int64(t.maxSlack):
			// a lot of nanoseconds passed since the last Take call
			// we will limit max accumulated time to maxSlack
			newTimeOfNextPermissionIssue = now - int64(t.maxSlack)
		default:
                 	// calculate the time at which our permission was issued
			newTimeOfNextPermissionIssue = timeOfNextPermissionIssue + int64(t.perRequest)
                 }
```

I think this is equivalent (?), and I like that we only have a single `CompareAndSwapInt64` - WDYT?

Separately, I was thinking about changing the initialization of the limiter so that it always inits with valid `now`. This way we don't have to handle the `0 case` in the "hot" path.

Reply:
I'll try to rewrite it in such way

Reply:
If you do, please just do the `switch`.

For the second part (initialization change) let's keep it separate - I'd like to keep the initialization consistent across all implementation. I was just asking for your opinion about that.

Reply:
@rabbbit 
Hey, I made the following [implementation in a separate branch](https://github.com/storozhukBM/ratelimit/blob/non_zero_state_and_time_based_calc/limiter_atomic_int64_with_non_zero_state.go) it is a bit simpler, but also a 3.3 ns slower, 27.4ns vs 24.1ns. I think it is because of using `time.Time.Sub()` or `time.Time.Add()`, that make some additional checks and handle internal state that `int64` math doesn't do.

<img width="1040" alt="Screenshot 2022-06-05 at 21 25 45" src="https://user-images.githubusercontent.com/3532750/172069499-a6ace080-1ef7-4265-bb8d-16f1e8d77e1d.png">


But this implementation handles zero state better. What do you think if I use that approach for state handling, but leave  int64 math around?

We can also merge this PR as is and deal with refactoring in subsequent PR.

Reply:
I think we're conflating 3 different changes here, right?

1. initial state change
2. single vs multiple CAS
3. int64 vs time types

I'm not okay with changing (1) (sorry, I left a note on this above too) - if we decide to change the initial state, we should change it in all implementations, and potentially add a test to formalize the startup behavior. I would definitely do this separately, or perhaps not at all.

I think I'd like us to decide on (2) in the current PR. As of my current understanding, there's no performance benefit in the original version, and the code seems clearer. So this seems like something we should just do.

I'm okay with leaving (3) as a follow up. Seems like a tradeoff between a tiny bit of speed (2ns-3ns) vs developer experience (compiler help). There's also a question of correctness, but I'm not right now sure how the serialization to int64 plays with the monotonic clock. We can potentially iterate on that separately.

Sorry for the amount of questions/comments :)

Reply:
I agree on all 3 points. I'll quickly double check if switch has OK perf and make this change. For 2 other things we can discuss later in issues maybe.

Reply:
Results are definitely strange, but [switch](https://github.com/storozhukBM/ratelimit/blob/non_zero_state_and_time_based_calc/limiter_atomic_int64_switch.go) is slower:

<img width="1047" alt="Screenshot 2022-06-06 at 00 29 47" src="https://user-images.githubusercontent.com/3532750/172075119-0ad9a7c7-190f-427d-8aad-01223723d324.png">

Reply:
I'm sorry about the number of comments. If this is getting annoying, let me know, and I can take over.

I'm okay with landing the switch version, and then iterating. For now:
- I prefer the readability over 5ns
- we're now significantly faster than the previous iteration anyway.

However, ^ made me realize is that I don't necessarily care about `switch` being present, but I'd like to have a single `CAS` operation. This means we could probably do something like:

```golang
		if timeOfNextPermissionIssue == 0 {
	                 // If this is our first request, then we allow it.	
			newTimeOfNextPermissionIssue = now - int64(t.perRequest)
                 }

		if now-timeOfNextPermissionIssue > int64(t.maxSlack) {
			// a lot of nanoseconds passed since the last Take call
			// we will limit max accumulated time to maxSlack
			newTimeOfNextPermissionIssue = now - int64(t.maxSlack)
                } else {
			// calculate the time at which our permission was issued
			newTimeOfNextPermissionIssue = timeOfNextPermissionIssue + int64(t.perRequest)
                }

		if atomic.CompareAndSwapInt64(&t.state, timeOfNextPermissionIssue, newTimeOfNextPermissionIssue) {
			break
		}
```

Which I think is largely equivalent to your original version, but easier to parse? We're doing exactly one more subtraction on the startup phase, but that shouldn't matter.

But, again, I'm okay with the switch version if you'd like to land that. And we can iterate from that.

Reply:
Yes, let's land switch for now and then iterate. It can be that only one CAS is causing the slowdown, or maybe we won't need handling of `timeOfNextPermissionIssue == 0` in the future. We will see.

Reply:
Done